File: 방꾸(E1I4)/Bangkoo-Backend/src/main/java/com/bangkoo/back/service/auth/UserService.java
package com.bangkoo.back.service.auth;

import com.bangkoo.back.model.auth.User;
import com.bangkoo.back.repository.auth.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.Optional;

@Service
@RequiredArgsConstructor
public class UserService {

    /**
     * 유저 저장과 조회 관련 서비스
     *
     */
    private final UserRepository userRepository;

    public User findOrCreate(String email, String nickname,
                             String role) {
        Optional<User> userOptional  = userRepository.findByEmail(email);

        return userOptional .orElseGet(() ->{
            User newUser = User.builder()
                    .email(email)
                    .nickname(nickname)
                    .role(role)
                    .build();
            return userRepository.save(newUser);
        });
    }
}

File: Bask-EAT((주)식회사)/Bask-EAT-Service-Backend/bin/main/com/bask_eat/business_backend/service/UserService.kt
package com.bask_eat.business_backend.service


import com.bask_eat.business_backend.model.entity.User
import com.bask_eat.business_backend.model.enums.UserRole
import com.google.firebase.auth.FirebaseToken
import org.springframework.stereotype.Service
import java.util.*

@Service
class UserService(
  private val firestoreService: FirestoreService
) {

  suspend fun findOrCreateUser(decodedToken: FirebaseToken): User {
    val uid = decodedToken.uid
    val email = decodedToken.email
    val name = decodedToken.name
    val picture = decodedToken.picture

    // 기존 사용자 조회
    var user = firestoreService.findUserById(uid)

    if (user == null) {
      // 새 사용자 생성
      user = User(
        id = uid,
        email = email,
        name = name ?: "사용자",
        profileImageUrl = picture,
        provider = "google",
        providerId = uid,
        role = UserRole.USER,
        createdAt = Date(),
        updatedAt = Date(),
        isActive = true
      )

      firestoreService.saveUser(user)
    } else {
      // 기존 사용자 정보 업데이트
      val updatedUser = user.copy(
        name = name ?: user.name,
        profileImageUrl = picture ?: user.profileImageUrl,
        updatedAt = Date()
      )

      firestoreService.saveUser(updatedUser)
      user = updatedUser
    }

    return user
  }

  suspend fun findById(userId: String): User? {
    return firestoreService.findUserById(userId)
  }

  suspend fun updateProfile(userId: String, updateRequest: Map<String, String>): User? {
    val user = findById(userId) ?: return null

    val updatedUser = user.copy(
      name = updateRequest["name"] ?: user.name,
      updatedAt = Date()
    )

    return firestoreService.saveUser(updatedUser)
  }
}
